File: backend/pkg/constants/fields.go
package constants

// System field names - standard fields present on all/most objects.
// These are the snake_case API names used in storage and SQL.
const (
	// Primary Fields defined in z_generated.go
	// FieldID, FieldName

	// Ownership and Security defined in z_generated.go
	// FieldOwnerID

	// Audit Fields defined in z_generated.go
	// FieldCreatedDate, FieldCreatedByID, FieldLastModifiedDate, FieldLastModifiedByID

	// Soft Delete defined in z_generated.go
	// FieldIsDeleted

	// User Fields
	FieldEmail         = "email"
	FieldUsername      = "username"
	FieldPassword      = "password"
	FieldFirstName     = "first_name"
	FieldLastName      = "last_name"
	FieldProfileID     = "profile_id"
	FieldRoleID        = "role_id"
	FieldIsActive      = "is_active"
	FieldLastLoginDate = "last_login_date"

	// Object/Field Metadata
	FieldAPIName     = "api_name"
	FieldLabel       = "label"
	FieldPluralLabel = "plural_label"
	FieldDescription = "description"
	FieldMetaType    = "type"
	FieldObjectID    = "object_id"
	FieldReferenceTo = "reference_to"
	FieldIsCustom    = "is_custom"
	FieldIsSystem    = "is_system"
	FieldIsRequired  = "required"
	FieldIsUnique    = "unique"

	// Flow/Action Fields
	FieldTriggerObject = "trigger_object"
	FieldTriggerType   = "trigger_type"
	FieldCondition     = "condition"
	FieldConfig        = "config"
	FieldStatus        = "status"
	FieldSortOrder     = "sort_order"

	// Recycle Bin Fields
	FieldRecordID      = "record_id"
	FieldObjectAPIName = "object_api_name"
	FieldRecordName    = "record_name"
	FieldDeletedBy     = "deleted_by"
	FieldDeletedDate   = "deleted_date"

	// Recent Items Fields
	FieldUserID    = "user_id"
	FieldTimestamp = "timestamp"

	// Config Fields
	FieldKeyName  = "key_name"
	FieldValue    = "value"
	FieldIsSecret = "is_secret"

	// Log Fields
	FieldLevel   = "level"
	FieldSource  = "source"
	FieldMessage = "message"
	FieldDetails = "details"

	// Session Fields
	FieldToken        = "token"
	FieldExpiresAt    = "expires_at"
	FieldLastActivity = "last_activity"
	FieldIPAddress    = "ip_address"
	FieldUserAgent    = "user_agent"
	FieldIsRevoked    = "is_revoked"

	// Metadata Fields
	FieldTableName     = "table_name"
	FieldTableType     = "table_type"
	FieldCategory      = "category"
	FieldIsManaged     = "is_managed"
	FieldSchemaVersion = "schema_version"
	FieldCreatedBy     = "created_by"
)

// StandardSystemFields returns the list of standard system field names
// that are present on most business objects.
func StandardSystemFields() []string {
	return []string{
		FieldID,
		FieldOwnerID,
		FieldCreatedDate,
		FieldCreatedByID,
		FieldLastModifiedDate,
		FieldLastModifiedByID,
		FieldIsDeleted,
	}
}

// IsSystemField checks if a field name is a standard system field
func IsSystemField(fieldName string) bool {
	for _, sf := range StandardSystemFields() {
		if sf == fieldName {
			return true
		}
	}
	return false
}

// AuditFields returns the audit field names
func AuditFields() []string {
	return []string{
		FieldCreatedDate,
		FieldCreatedByID,
		FieldLastModifiedDate,
		FieldLastModifiedByID,
	}
}

File: frontend/src/constants/fields.ts
/**
 * System Field Constants
 * 
 * Imports from shared JSON - single source of truth matching protobuf enums.
 * @see shared/constants/system.json
 * @see shared/proto/constants.proto (SystemField enum)
 */

import systemConstants from '@shared/constants/system.json';

// Extract system fields from shared JSON
const { systemFields, recordStates, defaults } = systemConstants;

// Field API names - derived from shared JSON
export const FieldID = systemFields.ID.apiName;
export const FieldName = systemFields.NAME.apiName;
export const FieldOwnerID = systemFields.OWNER_ID.apiName;
export const FieldCreatedDate = systemFields.CREATED_DATE.apiName;
export const FieldCreatedByID = systemFields.CREATED_BY_ID.apiName;
export const FieldLastModifiedDate = systemFields.LAST_MODIFIED_DATE.apiName;
export const FieldLastModifiedByID = systemFields.LAST_MODIFIED_BY_ID.apiName;
export const FieldIsDeleted = systemFields.IS_DELETED.apiName;



// Standard system fields array
export const SYSTEM_FIELDS = [
    FieldID,
    FieldOwnerID,
    FieldCreatedDate,
    FieldCreatedByID,
    FieldLastModifiedDate,
    FieldLastModifiedByID,
    FieldIsDeleted,
] as const;

// Audit fields array
export const AUDIT_FIELDS = [
    FieldCreatedDate,
    FieldCreatedByID,
    FieldLastModifiedDate,
    FieldLastModifiedByID,
] as const;

/**
 * Check if a field is a standard system field
 */
export function isSystemField(fieldName: string): boolean {
    return SYSTEM_FIELDS.includes(fieldName as typeof SYSTEM_FIELDS[number]);
}

/**
 * Check if a field is an audit field
 */
export function isAuditField(fieldName: string): boolean {
    return AUDIT_FIELDS.includes(fieldName as typeof AUDIT_FIELDS[number]);
}


